Solution for running/packaging native modules for Electron

While attempting to integrate my Electron app using `rules_node`, I've run into a roadblock with dependent modules that have a native build portion, in particular `grpc`.

At first I thought I could patch the call to `yarn` with some flag telling it to "build/install for Electron", which would result in a `yarn_modules` repository specifically built for Electron, but it seems this is not possible.

Do you have any ideas about how this could be implemented?

Hi @pcj, awesome that you're keen to support! I put together a minimal example [here](https://github.com/jondo2010/rules_node_grpc_electron)

With the sample output below, you can see that there is a mismatch with the gRPC binary module portion. It needs to be compiled against the correct version of Electron, and this is different than what yarn installs.

```
$ bazel run '//:test_bin'
INFO: Analysed target //:test_bin (2 packages loaded).
INFO: Found 1 target...
Target //:test_bin up-to-date:
  bazel-bin/test_bin
INFO: Elapsed time: 73.412s, Critical Path: 4.63s
INFO: Build completed successfully, 5922 total actions

INFO: Running command line: bazel-bin/test_bin
stderr: App threw an error during load

stderr: Error: Failed to load gRPC binary module because it was not installed for the current system
Expected directory: electron-v1.8-linux-x64-glibc
Found: [node-v59-linux-x64-glibc]
This problem can often be fixed by running "npm rebuild" on the current system
Original error: Cannot find module '/home/john/.cache/bazel/_bazel_john/8afb8bb0ba5daba4f60180e50fbdc72a/execroot/rules_node_grpc_electron/bazel-out/k8-fastbuild/bin/test_bin_files/node_modules/grpc/src/node/extension_binary/electron-v1.8-linux-x64-glibc/grpc_node.node'
    at Object.<anonymous> (/home/john/.cache/bazel/_bazel_john/8afb8bb0ba5daba4f60180e50fbdc72a/execroot/rules_node_grpc_electron/bazel-out/k8-fastbuild/bin/test_bin_files/node_modules/grpc/src/grpc_extension.js:44:17)
    at Object.<anonymous> (/home/john/.cache/bazel/_bazel_john/8afb8bb0ba5daba4f60180e50fbdc72a/execroot/rules_node_grpc_electron/bazel-out/k8-fastbuild/bin/test_bin_files/node_modules/grpc/src/grpc_extension.js:54:3)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
    at Module.require (module.js:513:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/john/.cache/bazel/_bazel_john/8afb8bb0ba5daba4f60180e50fbdc72a/execroot/rules_node_grpc_electron/bazel-out/k8-fastbuild/bin/test_bin_files/node_modules/grpc/src/client.js:38:12)
A JavaScript error occurred in the main process
Uncaught Exception:
Error: Failed to load gRPC binary module because it was not installed for the current system
Expected directory: electron-v1.8-linux-x64-glibc
Found: [node-v59-linux-x64-glibc]
This problem can often be fixed by running "npm rebuild" on the current system
Original error: Cannot find module '/home/john/.cache/bazel/_bazel_john/8afb8bb0ba5daba4f60180e50fbdc72a/execroot/rules_node_grpc_electron/bazel-out/k8-fastbuild/bin/test_bin_files/node_modules/grpc/src/node/extension_binary/electron-v1.8-linux-x64-glibc/grpc_node.node'
    at Object.<anonymous> (/home/john/.cache/bazel/_bazel_john/8afb8bb0ba5daba4f60180e50fbdc72a/execroot/rules_node_grpc_electron/bazel-out/k8-fastbuild/bin/test_bin_files/node_modules/grpc/src/grpc_extension.js:44:17)
    at Object.<anonymous> (/home/john/.cache/bazel/_bazel_john/8afb8bb0ba5daba4f60180e50fbdc72a/execroot/rules_node_grpc_electron/bazel-out/k8-fastbuild/bin/test_bin_files/node_modules/grpc/src/grpc_extension.js:54:3)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
    at Module.require (module.js:513:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/john/.cache/bazel/_bazel_john/8afb8bb0ba5daba4f60180e50fbdc72a/execroot/rules_node_grpc_electron/bazel-out/k8-fastbuild/bin/test_bin_files/node_modules/grpc/src/client.js:38:12)
```

--- a/README.md
+++ b/README.md
@@ -114,6 +114,7 @@ node_module(
 | --- | --- | --- | --- |
 | optional | `label` | `package_json` | A `package.json` file containing the dependencies that should be installed. |
 | optional | `string_dict` | `deps` | A mapping of `name` --> `version` for the dependencies that should be installed. |
+| optional | `string_list` | `post_install` | A list of command-line arguments that should be invoked after the `yarn install` step.  See [#55](https://github.com/pubref/rules_node/issues/55). |
 
 > Either `package_json` or `deps` must be present, but not both.
 

--- a/node/internal/clean_node_modules.js
+++ b/node/internal/clean_node_modules.js
@@ -1,0 +1,23 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path')
+
+// Yes, could look for BUILD files also, but wait for that error to
+// occur before fixing it.
+visitSync('node_modules/', filename => {
+  if (filename.endsWith("BUILD.bazel")) {
+    fs.unlink(filename);
+    console.log("Removed", filename, "from the node_modules/ tree");
+  }
+});
+
+function visitSync(d, visitFn) {
+  if (fs.statSync(d).isDirectory()) {
+    fs.readdirSync(d).forEach(name => visitSync(path.join(d, name), visitFn));
+  } else {
+    visitFn(d);
+  }                             
+}
+          
+          

--- a/node/internal/yarn_modules.bzl
+++ b/node/internal/yarn_modules.bzl
@@ -32,6 +32,7 @@ def _yarn_modules_impl(ctx):
     node = ctx.path(node_label)
 
     parse_yarn_lock_js = ctx.path(ctx.attr._parse_yarn_lock_js)
+    clean_node_modules_js = ctx.path(ctx.attr._clean_node_modules_js)
     yarn_js = ctx.path(ctx.attr._yarn_js)
 
 
@@ -53,6 +54,7 @@ def _yarn_modules_impl(ctx):
         
     # Copy the parse_yarn_lock script and yarn.js over here.
     execute(ctx, ["cp", parse_yarn_lock_js, "internal/parse_yarn_lock.js"])
+    execute(ctx, ["cp", clean_node_modules_js, "internal/clean_node_modules.js"])
     execute(ctx, ["cp", yarn_js, "yarn.js"])
 
     # Build node_modules via 'yarn install'
@@ -61,9 +63,32 @@ def _yarn_modules_impl(ctx):
         "PATH": "%s:$PATH" % ctx.path(node).dirname,
     })
 
+    # Sadly, pre-existing BUILD.bazel files located in npm modules can
+    # screw up package boudaries.  Remove these.
+    execute(ctx, [node, "internal/clean_node_modules.js"], quiet = True)
+    
     # Run the script and save the stdout to our BUILD file(s)
     parse_args = ["--resolve=%s:%s" % (k, v) for k, v in ctx.attr.resolutions.items()]
     result = execute(ctx, [node, "internal/parse_yarn_lock.js"] + parse_args, quiet = True)
+
+    # If the user has specified a post_install step, build those args
+    # and execute it.
+    post_install_args = []
+    for arg in ctx.attr.post_install:
+        if arg == "{node}":
+            post_install_args.append(node)
+        else:
+            post_install_args.append(arg)
+    if len(post_install_args) > 0:
+        # Expose python, node, and basic tools by default.  Might need
+        # to add a post_install_tools option in the future if this is
+        # not sufficient.
+        python = ctx.which("python")
+        mkdir = ctx.which("mkdir")
+        execute(ctx, post_install_args, environment = {
+            "PATH": ":".join([node.dirname, python.dirname, mkdir.dirname, "$PATH"])
+        })
+
     ctx.file("BUILD.bazel", result.stdout)
 
 
@@ -89,6 +114,10 @@ yarn_modules = repository_rule(
             default = Label("//node:internal/parse_yarn_lock.js"),
             single_file = True,
         ),
+        "_clean_node_modules_js": attr.label(
+            default = Label("//node:internal/clean_node_modules.js"),
+            single_file = True,
+        ),
         "_yarn_js": attr.label(
             default = Label("@yarn//:bin/yarn.js"),
             single_file = True,
@@ -97,6 +126,7 @@ yarn_modules = repository_rule(
             mandatory = False,
             allow_files = FileType(["package.json"]),
         ),
+        "post_install": attr.string_list(),
         "deps": attr.string_dict(mandatory = False),
         "resolutions": attr.string_dict(mandatory = False),
     }

